File: 算法训练营/06-树/144/递归法.py
#
# @lc app=leetcode.cn id=144 lang=python3
#
# [144] 二叉树的前序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        ret = []
        self.helper(root, ret)
        return ret
    
    def helper(self, root, ret):
        if root:
            ret.append(root.val)
            self.helper(root.left, ret)
            self.helper(root.right, ret)
# @lc code=end

File: 算法训练营/08-分治回溯/050/递归法.go
/*
 * @lc app=leetcode.cn id=50 lang=golang
 *
 * [50] Pow(x, n)
 */

// @lc code=start
func myPow(x float64, n int) float64 {
	switch {
	case x == 0.0:
		return 0.0
	case n == 0:
		return 1.0
	}

	return helper(x, n)
}

func helper(x float64, n int) float64 {
	if n < 0 {
		return helper(1/x, -n)
	} else if n == 0 {
		return 1.0
	}

	if n % 2 == 1 {
		return x * helper(x, n-1)
	}
	return helper(x*x, n>>1)
}
// @lc code=end